npm/private/npm_link_package.bzl

```diff
@@ -289,22 +290,30 @@ deps of npm_link_package_store must be in the same package.""" % (ctx.label.pack
                         for dep_ref_dep_alias in dep_ref_dep_aliases:
                             # "node_modules/{virtual_store_root}/{virtual_store_name}/node_modules/{package}"
                             dep_ref_dep_symlink_path = paths.join("node_modules", utils.virtual_store_root, dep_virtual_store_name, "node_modules", dep_ref_dep_alias)
-                            direct_files.extend(_make_symlink(ctx, dep_ref_dep_symlink_path, dep_ref_def_virtual_store_directory))
+                            transitive_files.extend(_make_symlink(ctx, dep_ref_dep_symlink_path, dep_ref_def_virtual_store_directory))
 
     direct_files = depset(direct = direct_files)
-    files_depsets = [direct_files]
-    runfiles = ctx.runfiles(transitive_files = direct_files)
+    transitive_files = depset(direct = transitive_files)
+    transitive_closure_depsets = [direct_files, transitive_files]
+    transitive_declarations_depsets = []
+
+    runfiles = ctx.runfiles(transitive_files = depset(transitive = [direct_files, transitive_files]))
+
     for dep in ctx.attr.deps:
-        files_depsets.append(dep[DefaultInfo].files)
+        transitive_closure_depsets.append(dep[DefaultInfo].files)
+        if DeclarationInfo in dep:
+            transitive_declarations_depsets.append(dep[DeclarationInfo].transitive_declarations)
         runfiles = runfiles.merge(dep[DefaultInfo].data_runfiles)
 
     result = [
-        DefaultInfo(files = depset(transitive = files_depsets), runfiles = runfiles),
+        DefaultInfo(files = depset(transitive = transitive_closure_depsets), runfiles = runfiles),
         # Always assume that packages provide typings, so we don't need to use an action to
-        # inspect the package.json#typings field or search for .d.ts files in the package.
-        declaration_info(
+        # inspect the package.json#typings field or search for .d.ts files in the package and
+        # so that any package can be passed as dep to rules that require DeclarationInfo such
+        # as ts_project.
+        DeclarationInfo(
             declarations = direct_files,
-            deps = ctx.attr.deps.keys(),
+            transitive_declarations = depset(transitive = [transitive_files] + transitive_declarations_depsets),
```

Comment:
Why are `transitive_files` in `transitive_declarations`? Is that just a naming thing?

Reply:
It's a convention through rule sets. `foo` typically are direct foo things in a target and `transtive_foo` the direct foo things plus all the foo things from all deps & data.

Similar to `transitive_sources` in `PyInfo`L https://github.com/aspect-build/rules_py/blob/main/py/private/py_library.bzl#L98

Reply:
But we're putting "files" into "declarations". Shouldn't declarations only be declaration files?

Reply:
This is out of date now that the code has been refactored further